xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/internal/html/SecureHTMLElementSanitizer.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.xml.internal.html;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.stability.Unstable;
+import org.xwiki.xml.html.HTMLElementSanitizer;
+
+/**
+ * Secure default implementation of {@link HTMLElementSanitizer} based on a definition of allowed elements and
+ * attributes.
+ * <p>
+ * This is heavily inspired by DOMPurify by Cure53 and other contributors | Released under the Apache license 2.0 and
+ * Mozilla Public License 2.0 - <a href="https://github.com/cure53/DOMPurify/blob/main/LICENSE">LICENSE</a>.
+ *
+ * @version $Id$
+ * @since 14.4RC1
+ * @since 13.10.6
+ */
+@Component
+@Named(SecureHTMLElementSanitizer.HINT)
+@Singleton
+@Unstable
+public class SecureHTMLElementSanitizer implements HTMLElementSanitizer
+{
+    /**
+     * The hint of this component.
+     */
+    public static final String HINT = "secure";
+
+    static final Pattern IS_SCRIPT_OR_DATA = Pattern.compile("^(?:\\w+script|data):", Pattern.CASE_INSENSITIVE);
+
+    static final Pattern ATTR_WHITESPACE =
+        Pattern.compile("[\\u0000-\\u0020\\u00A0\\u1680\\u180E\\u2000-\\u2029\\u205F\\u3000]");
+
+    static final Pattern DATA_ATTR = Pattern.compile("^data-[\\-\\w.\\u00B7-\\uFFFF]");
+
+    static final Pattern ARIA_ATTR = Pattern.compile("^aria-[\\-\\w]+$");
+
+    /**
+     * Allowed HTML elements.
+     */
+    private final Set<String> htmlTags;
+
+    /**
+     * Allowed attributes.
+     */
+    private final Set<String> htmlAttributes;
+
+    /**
+     * Tags that are safe for data: URIs.
+     */
+    private final Set<String> dataUriTags;
+
+    /**
+     * Attributes safe for values like "javascript:".
+     */
+    private final Set<String> uriSafeAttributes;
+
+    /**
+     * Default constructor.
+     */
+    public SecureHTMLElementSanitizer()
+    {
+        this.htmlTags = new HashSet<>(
```

Comment:
I have the feeling that we should make this setup customizable through configuration, it would make easier for someone to workaround the fact that we are missing a safe element (new one or simply a mistake) for example.

Reply:
I've actually wondered about this too. What do you suggest for the configuration variables, a comma-separated list of (fixed) names or a regular expression?

Reply:
I was thinking about simple lists (xwiki-properties already has its own syntaxes for lists, so you can directly ask for one, no need to parse a string) but indeed you have a good point with the regex expression, there might be use cases where it's needed.

Reply:
I would propose to first implement something similar to what the upstream source does. There, the configured allowed tags and attributes replace the default ones. I would propose to instead add extra tags and attributes via the configuration. Similar to the upstream, I would also add options to:
* Specify additional URI-safe attributes (i.e., attributes where arbitrary content should be allowed).
* Specify additional attributes where data-URIs shall be allowed.
* Forbid tags.
* Forbid attributes.

Should we also add the option to restrict URIs to allowed URIs with the option to override the [default regular expression](https://github.com/cure53/DOMPurify/blob/0965e9951dbd5bf79fee922dda2fc59f36ade0e7/src/regexp.js#L9)? I've so far just implemented the variant that allows arbitrary protocols (but forbids scripts and data-URIs) as it seemed quite restrictive to me and I would bet that this is the first thing where users complain that they cannot use their custom URIs to open their application. Upstream has URI restrictions as default and provides the option to allow all protocols, we could also reverse the default.